akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ClusterActorSetImpl.scala

```diff
@@ -0,0 +1,178 @@
+              startChild()
+              Behaviors.same
+          }
+      }
+
+    private def waitingForChildToStop(): Behavior[Command] = Behaviors.setup { context =>
+      // FIXME timeout and force termination if child doesn't stop?
+      Behaviors.receiveSignal {
+        case (_, Terminated(_)) =>
+          // if the child stopped manually, but we are not being re-located to another node,
+          // we can fast-restart it instead of waiting for a ping
+          context.log.debug(s"ClusterActorSet actor stopped")
+          Behaviors.stopped
+      }
+    }
+  }
+
+  object KeepAlivePinger {
+    sealed trait Event
+    case object Tick extends Event
+
+    def apply(
+        settings: ClusterActorSetSettings,
+        identities: Set[EntityId],
+        shardingRef: ActorRef[ShardingEnvelope[EntityParent.Command]]): Behavior[Event] =
+      Behaviors.setup { context =>
+        Behaviors.withTimers { timers =>
+          context.log.debug(
+            s"Starting ClusterActorSet KeepAlivePinger with ping interval ${settings.keepAliveInterval}")
+          // FIXME should we delay initial tick with random fraction of keepAliveInterval to avoid thundering herd?
+          timers.startTimerWithFixedDelay(Tick, settings.keepAliveInterval)
+          Behaviors.receiveMessage {
+            case Tick =>
+              identities.foreach(id => shardingRef ! ShardingEnvelope(id, EntityParent.Ping))
+              Behaviors.same
+          }
+        }
+      }
+  }
+
+}
+@InternalApi
+private[akka] class ClusterActorSetImpl(system: ActorSystem[_]) extends ClusterActorSet {
+
+  private val typeKeyCounter = new AtomicInteger(0)
+
+  override def init[T](numberOfEntities: Int, behaviorFactory: EntityId => Behavior[T], stopMessage: T): Unit =
+    init(ClusterActorSetSettings(system), numberOfEntities, behaviorFactory, stopMessage)
+
+  override def init[T](
+      settings: ClusterActorSetSettings,
+      numberOfEntities: Int,
+      behaviorFactory: EntityId => Behavior[T],
+      stopMessage: T): Unit = {
+    val identities = (0 to numberOfEntities).map(_.toString).toSet
+    init(settings, identities, behaviorFactory, stopMessage)
+  }
+
+  override def init[T](identities: Set[EntityId], behaviorFactory: EntityId => Behavior[T], stopMessage: T): Unit =
+    init(ClusterActorSetSettings(system), identities, behaviorFactory, stopMessage)
+
+  override def init[T](
+      settings: ClusterActorSetSettings,
+      identities: Set[EntityId],
+      behaviorFactory: EntityId => Behavior[T],
+      stopMessage: T): Unit = {
+    val setId = typeKeyCounter.incrementAndGet()
+    val entityTypeKey = EntityTypeKey[EntityParent.Command](s"cluster-actor-set-$setId")
+
+    // Since we know up front exactly what entity ids will exist and the number will be low 1:1 is fine as it will
+    // balance the set as good as possible across the cluster
+    val numberOfShards = identities.size
```

Comment:
👍 do we document some where that this is only for lowish nrs?

Reply:
we'll need a message extractor to set the shard id to entity id or they may end up in the same shard

Reply:
+1 extractor that is 1:1 mapping entityId-shardId